crates/astria-merkle/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+///
+/// The proof is the concatenation of all merkle tree hashes from
+/// a node to the root.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Proof {
+    audit_path: Vec<u8>,
+    leaf_index: usize,
+    tree_size: usize,
+}
+
+impl Proof {
+    /// Returns the audit path of the proof.
+    ///
+    /// This is the concatenation of all hashes to reconstruct merkle tree root
+    /// from the node under proof.
+    #[must_use]
+    #[inline]
+    pub fn audit_path(&self) -> &[u8] {
+        &self.audit_path
+    }
+
+    /// Returns the leaf index
+    #[must_use]
+    #[inline]
+    pub fn leaf_index(&self) -> usize {
+        self.leaf_index
+    }
+
+    /// Returns if the proof is empty.
+    ///
+    /// This can happen if the proof was constructed for a tree containing only
+    /// one leaf.
+    #[must_use]
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.audit_path.is_empty()
+    }
+
+    /// Returns the number of segments in the proof.
+    #[must_use]
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.audit_path.len() / 32
+    }
+
+    /// Returns the size of the tree this proof was derived from.
+    #[must_use]
+    #[inline]
+    pub fn tree_size(&self) -> usize {
+        self.tree_size
+    }
+
+    /// Walks the audit path to reconstruct the root hash starting from a node hash.
+    ///
+    /// Use this method if the starting point is the hash of a node (like the leaf hash)
+    /// which does not need to be hashed prior to combining with its sibling.
+    ///
+    /// # Examples
+    /// ```
+    /// # use astria_merkle::Tree;
+    /// let mut tree = Tree::new();
+    /// tree.push(&[1; 32]);
+    /// tree.push(&[2; 32]);
+    /// tree.push(&[3; 32]);
+    /// tree.push(&[4; 32]);
+    /// tree.push(&[5; 32]);
+    /// tree.push(&[6; 32]);
+    /// tree.push(&[7; 32]);
+    /// tree.push(&[8; 32]);
+    /// let root = tree.root();
+    /// let proof = tree
+    ///     .construct_proof(6)
+    ///     .expect("leaf 7 must be inside the tree");
+    /// let leaf_hash = tree.leaf_hash(6).expect("leaf 7 must be inside the tree");
+    /// let reconstructed_root = proof.reconstruct_root_with_leaf_hash(leaf_hash);
+    /// assert_eq!(root, reconstructed_root);
+    /// ```
+    #[must_use]
+    pub fn reconstruct_root_with_leaf_hash(&self, leaf_hash: [u8; 32]) -> [u8; 32] {
+        let Self {
+            audit_path,
+            leaf_index,
+            tree_size,
+        } = self;
+        let mut i = leaf_index_to_tree_index(*leaf_index);
+        let mut acc = leaf_hash;
+        for sibling in audit_path.chunks(32) {
+            let parent = complete_parent(i, *tree_size);
+            if parent > i {
+                acc = combine(&acc, sibling);
+            } else {
+                acc = combine(sibling, &acc);
+            }
+            i = parent;
+        }
+        acc
```

Comment:
could there be an error in the case where the proof is too long/too short? should there be a check that the proof is the expected length?